src/main/java/com/quorum/gauge/services/TransactionService.java

```diff
@@ -172,4 +176,135 @@
+        Web3j client = connectionFactory.getWeb3jConnection(from);
+        String fromAddress;
+        try {
+            fromAddress = client.ethCoinbase().send().getAddress();
+        } catch (IOException e) {
+            logger.error("Unable to get default account for node " + from, e);
+            throw new RuntimeException(e);
+        }
+
+        String data = c.getContractBinary();
+        return client.ethGetTransactionCount(fromAddress, DefaultBlockParameterName.LATEST)
+                .observable()
+                .flatMap(ethGetTransactionCount -> {
+                    Transaction tx = Transaction.createContractTransaction(fromAddress,
+                            ethGetTransactionCount.getTransactionCount(),
+                            BigInteger.valueOf(0),
+                            DEFAULT_GAS_LIMIT,
+                            BigInteger.valueOf(0),
+                            data);
+                    return client.ethEstimateGas(tx).observable();
+                });
+    }
+
+    public Observable<EthEstimateGas> estimateGasForPrivateContract(QuorumNode from, QuorumNode privateFor, Contract c) {
+        Web3j client = connectionFactory.getWeb3jConnection(from);
+        String fromAddress;
+        try {
+            fromAddress = client.ethCoinbase().send().getAddress();
+        } catch (IOException e) {
+            logger.error("Unable to get default account for node " + from, e);
+            throw new RuntimeException(e);
+        }
+
+        String data = c.getContractBinary();
+        return client.ethGetTransactionCount(fromAddress, DefaultBlockParameterName.LATEST)
+                .observable()
+                .flatMap(ethGetTransactionCount -> {
+                    Transaction tx = new PrivateTransaction(
+                            fromAddress,
+                            ethGetTransactionCount.getTransactionCount(),
+                            DEFAULT_GAS_LIMIT,
+                            null,
+                            BigInteger.valueOf(0),
+                            data,
+                            null,
+                            Arrays.asList(privacyService.id(privateFor))
+                    );
+                    return client.ethEstimateGas(tx).observable();
+                });
+    }
+
+    public Observable<EthEstimateGas> estimateGasForPublicContractCall(QuorumNode from, Contract c) {
+        Web3j client = connectionFactory.getWeb3jConnection(from);
+        String fromAddress;
+        try {
+            fromAddress = client.ethCoinbase().send().getAddress();
+        } catch (IOException e) {
+            logger.error("Unable to get default account for node " + from, e);
+            throw new RuntimeException(e);
+        }
+
+        //TODO: need to work out how to generate the data for SimpleContract.set(), instead of using hardcoded data
+        String data = "60fe47b10000000000000000000000000000000000000000000000000000000000000063";
```

Comment:
This can likely be done
```
Function function = new Function(
                FUNC_SET, 
                Arrays.<Type>asList(new org.web3j.abi.datatypes.generated.Uint256(x)), 
                Collections.<TypeReference<?>>emptyList());
String data = FunctionEncoder.encode(function);
```
I have not tested this but i traced back the logic from the generated code in `SimpleStorage.set()`

Reply:
Thanks - I updated my acceptance tests and that works a treat.